File: p2-regression/multiple-linear-regression/multiple-linear-regression.r
# multiple linear regression
# F1 to see docs

# import dataset
dataset = read.csv('50_Startups.csv')
# dataset = dataset[, 2:3]

# encode categorical data
dataset$State = factor(dataset$State,
                         level = c('New York', 'California', 'Florida'),
                         labels = c(1, 2 ,3))

# split dataset into training set and test set
# install.packages('caTools')
library(caTools)
set.seed(123)
# 20% goes into test
split = sample.split(dataset$Profit, SplitRatio = 0.8)
training_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)

# # feature scaling
# training_set[, 2:3] = scale(training_set[, 2:3])
# test_set[, 2:3] = scale(test_set[, 2:3])

# fit multiple linear regression on training set
# regressor = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend + State,
# data = training_set)
# . means all independent variables
regressor = lm(formula = Profit ~ .,
               data = training_set)

# predict test set results
y_pred = predict(regressor, newdata=test_set)

# build optimal model using backward elimination
regressor = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend + State,
               data = dataset)
summary(regressor)

# remove predictor - State
regressor = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend,
               data = dataset)
summary(regressor)

# remove Administration
regressor = lm(formula = Profit ~ R.D.Spend + Marketing.Spend,
               data = dataset)
summary(regressor)

# remove Marketing.Spend because > 0.05 p-value
regressor = lm(formula = Profit ~ R.D.Spend,
               data = dataset)
summary(regressor)

File: p2-regression/multiple-linear-regression/multiple-linear-regression.py
# multiple linear regression

# import libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# print full numpy array without truncation
np.set_printoptions(threshold=np.inf)

# import dataset
dataset = pd.read_csv('50_Startups.csv')
X = dataset.iloc[:, :-1].values
y = dataset.iloc[:, 4].values

# encode categorical data using dummy variables
# encode by independent variable
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
labelencoder_X = LabelEncoder()
X[:, 3] = labelencoder_X.fit_transform(X[:, 3])
onehotencoder = OneHotEncoder(categorical_features=[3])
X = onehotencoder.fit_transform(X).toarray()

# avoid dummy variable trap by removing one dummy variable
X = X[:, 1:]

# split dataset into training set and test set
#from sklearn.cross_validation import train_test_split
from sklearn.model_selection import train_test_split
# 20% goes into test set
X_train, X_test, y_train, y_test=train_test_split(X, y, test_size=0.2, random_state=0)

'''
# feature scaling
from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)
'''

# fit multiple linear regression to training set
from sklearn.linear_model import LinearRegression
regressor = LinearRegression()
regressor.fit(X_train, y_train)

# predict test set results
y_pred = regressor.predict(X_test)

# build optimal model using backward elimination
import statsmodels.formula.api as sm
X = np.append(arr=np.ones((50, 1)).astype(int), values=X, axis=1)
X_opt = X[:, [0,1,2,3,4,5]]
regressor_OLS = sm.OLS(endog=y, exog=X_opt).fit()
regressor_OLS.summary()

# remove index with highest P value (index 2)
X_opt = X[:, [0,1,3,4,5]]
regressor_OLS = sm.OLS(endog=y, exog=X_opt).fit()
regressor_OLS.summary()

# remove index 1
X_opt = X[:, [0,3,4,5]]
regressor_OLS = sm.OLS(endog=y, exog=X_opt).fit()
regressor_OLS.summary()

# remove index 4
X_opt = X[:, [0,3,5]]
regressor_OLS = sm.OLS(endog=y, exog=X_opt).fit()
regressor_OLS.summary()

# remove index 5
X_opt = X[:, [0,3]]
regressor_OLS = sm.OLS(endog=y, exog=X_opt).fit()
regressor_OLS.summary()



























